### PR TITLE
Fix kldiv backward on CUDA

### LINKED ISSUE #5801: KLDivLoss behaves differently on CPU/GPU
Pytorch Version: '0.3.0'

Code:
```
import torch
import torch.nn as nn
from torch.autograd import Variable
import sys

batchsize = 8
nclass = 5

if sys.argv[1] == "cuda":
    cuda = True
else:
    cuda = False

print("Cuda: " + str(cuda))

def make_prob():
    if cuda:
        pr = torch.cuda.FloatTensor(batchsize, nclass)
    else:
        pr = torch.FloatTensor(batchsize, nclass)
    pr.uniform_()
    pr /= pr.sum(dim=1).view(batchsize, 1)
    return pr

def loss_kl(pi, targets):
    pi_var = Variable(pi, requires_grad=True)
    logsoftmax = nn.LogSoftmax(dim=1)
    kl = nn.KLDivLoss()

    if cuda:
        logsoftmax.cuda()
        kl.cuda()

    logpi = logsoftmax(pi_var)
    return kl(logpi, Variable(targets)) * logpi.size(1), pi_var

def loss_plain(pi, targets):
    pi_var = Variable(pi, requires_grad=True)
    targets_var = Variable(targets)
    logsoftmax = nn.LogSoftmax(dim=1)

    if cuda:
        logsoftmax.cuda()

    logpi = logsoftmax(pi_var)
    loss = - (logpi * targets_var).sum(dim=1).mean() + (targets_var * targets_var.log()).sum(dim=1).mean()
    return loss, pi_var

targets = make_prob()
pi = make_prob()

loss1, pi_var1 = loss_kl(pi, targets)
print(loss1)
loss1.backward()
print(pi_var1.grad)

loss2, pi_var2 = loss_plain(pi, targets)
print(loss2)
loss2.backward()
print(pi_var2.grad)
```

Without cuda, both the gradients and the loss are the same:
```
$ python3 ~/test_kl.py nocuda
Cuda: False
Variable containing:
 0.1930
[torch.FloatTensor of size 1]

Variable containing:
1.00000e-02 *
  1.9849 -1.7913 -0.1536  0.1435 -0.1834
 -2.1244  2.5148  0.9348  0.5213 -1.8464
  0.3656  0.1773 -0.1752 -0.1085 -0.2591
 -1.7945 -0.2471  1.9382  1.0904 -0.9869
  1.6413  2.5043 -1.3458 -3.3194  0.5196
  0.0206 -1.2024  0.4363  0.8926 -0.1471
 -0.4564  2.5605  0.2420 -1.0382 -1.3080
  1.9253  1.7758 -1.0304 -0.5174 -2.1532
[torch.FloatTensor of size 8x5]

Variable containing:
 0.1930
[torch.FloatTensor of size 1]

Variable containing:
1.00000e-02 *
  1.9849 -1.7913 -0.1536  0.1435 -0.1834
 -2.1244  2.5148  0.9348  0.5213 -1.8464
  0.3656  0.1773 -0.1752 -0.1085 -0.2591
 -1.7945 -0.2471  1.9382  1.0904 -0.9869
  1.6413  2.5043 -1.3458 -3.3194  0.5196
  0.0206 -1.2024  0.4363  0.8926 -0.1471
 -0.4564  2.5605  0.2420 -1.0382 -1.3080
  1.9253  1.7758 -1.0304 -0.5174 -2.1532
[torch.FloatTensor of size 8x5]
```

With cuda the gradients returned by KLDivLoss are scaled down by the number of class, while the loss is the same. 
```
$ python3 ~/test_kl.py cuda
Cuda: True
Variable containing:
 0.1756
[torch.cuda.FloatTensor of size 1 (GPU 0)]

Variable containing:
1.00000e-03 *
 -1.2566 -3.5917 -0.8016  2.9373  2.7126
 -0.0264  2.3321 -1.3892  1.0676 -1.9841
  0.1204  4.5507 -2.2642 -3.7239  1.3170
  0.5667 -2.7765  4.3564 -4.8917  2.7452
  2.3703 -2.4578  2.7945  1.8861 -4.5931
 -0.2237 -0.2954  1.7677  1.7757 -3.0243
 -3.9952 -7.2011  2.1794  3.3324  5.6845
 -0.2814  0.0368 -2.2764 -0.2955  2.8164
[torch.cuda.FloatTensor of size 8x5 (GPU 0)]

Variable containing:
 0.1756
[torch.cuda.FloatTensor of size 1 (GPU 0)]

Variable containing:
1.00000e-02 *
 -0.6283 -1.7959 -0.4008  1.4686  1.3563
 -0.0132  1.1661 -0.6946  0.5338 -0.9920
  0.0602  2.2753 -1.1321 -1.8619  0.6585
  0.2833 -1.3883  2.1782 -2.4459  1.3726
  1.1852 -1.2289  1.3972  0.9430 -2.2966
 -0.1119 -0.1477  0.8838  0.8879 -1.5121
 -1.9976 -3.6006  1.0897  1.6662  2.8423
 -0.1407  0.0184 -1.1382 -0.1477  1.4082
[torch.cuda.FloatTensor of size 8x5 (GPU 0)]
```

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #5801 

Two fixes:
1) Fix incorrect kldiv backward on CUDA
2) Change Criterion testing to test that the criterion uses gradOutput. The problem with KLDiv is that it was not using gradOutput but we didn't have a test for this.

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -333,13 +333,13 @@ def _forward_criterion(self, criterion, input, target):
             output = criterion(input, target)
         return output.item()
 
-    def _backward_criterion(self, criterion, input, target):
+    def _backward_criterion(self, criterion, input, target, gradOutput=torch.ones(())):
```
Comment (colesbury) [inline test/test_nn.py]: Don't use mutable data structures as default args. The same value is used for every invocation.

Use `_backward_criterion(..., grad_output=None)` and then in the body:

```python
if grad_output is None:
   grad_output = torch.ones(())
````

Comment (zou3519) [reply test/test_nn.py]: Yikes I didn't know that. Good catch!

**Comment by SsnL** [comment]
you might want to rebase to make onnx pass..

**Comment by zou3519** [comment]
😢
